{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8cf80366",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Bio/pairwise2.py:283: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  BiopythonDeprecationWarning,\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import gzip\n",
    "#import pyfrost\n",
    "import numpy\n",
    "from collections import defaultdict\n",
    "import matplotlib.pyplot as plot\n",
    "#import CCGG_extension as CCGG\n",
    "import json\n",
    "import AGG\n",
    "import pandas as pd\n",
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq\n",
    "import Levenshtein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a2e55326",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = './MT_anchor_graphical_genome_k31_anno_delunspported.gfa'\n",
    "filename = './MT_anchor_graphical_genome_k31_anno_series_parallele.gfa'\n",
    "spgraph = AGG.GraphicalGenome(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "782388f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52=\n"
     ]
    }
   ],
   "source": [
    "edgelist = spgraph.edges.keys()\n",
    "for edge in edgelist:\n",
    "    print(spgraph.edges[edge].get('variants', \"\"))\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6721a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "#help(AGG.SubGraph)\n",
    "def find_sample_list(graph):\n",
    "    samplelist = []\n",
    "    edgelist = graph.edges.keys()\n",
    "    for edge in edgelist:\n",
    "        samplelist += graph.edges[edge]['strain']\n",
    "    samplelist = list(set(samplelist))\n",
    "    return samplelist\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9bc8dca6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#samplelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "edc91ecd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#'A000069' in subgraph.outgoing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2833a5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/suhang/Analysis/Mitochondrial/MitoHifi/Mitochondrial/NC_012920.1.fasta') as fp:\n",
    "    data = fp.readlines()\n",
    "data.pop(0)\n",
    "contig = \"\".join([l[:-1] for l in data])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "92912f2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample = '012920'\n",
    "subgraph = AGG.SubGraph(spgraph, [sample,  '012920'])\n",
    "path = AGG.find_most_supported_path(subgraph, [sample])\n",
    "p = path[0]\n",
    "contig.find(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "790cc7f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sorted(subgraph.anchor.keys())\n",
    "outputfile = \"test.fasta\"\n",
    "def write_fasta(outputfile, sequence, sample):\n",
    "    header = \">\" + sample + \", mitochondrial most supported path\" + '\\n'\n",
    "    with open(outputfile, 'w') as fp:\n",
    "        fp.write(header)\n",
    "        l = len(sequence)//60\n",
    "        for i in range(l+1):\n",
    "            seq = sequence[i*60:(i+1)*60]\n",
    "            fp.write(seq + '\\n')\n",
    "        fp.write(sequence[(i+1)*60:] + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1e5009e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NA18906 16651\n",
      "HG03453 16653\n",
      "HG00735 16651\n",
      "HG02109 16653\n",
      "012920 16407\n",
      "HG02572 16654\n",
      "HG02257 16653\n",
      "HG02717 16651\n",
      "HG03540 16650\n",
      "HG00733 16651\n",
      "HG01123 16651\n",
      "HG02886 16651\n",
      "HG01071 16651\n",
      "HG002 16652\n",
      "HG01243 16650\n",
      "HG02559 16652\n",
      "NA20129 16651\n",
      "HG01928 16652\n",
      "HG03098 16653\n",
      "HG02622 16651\n",
      "HG02080 16652\n",
      "HG01891 16650\n",
      "HG01175 16653\n",
      "HG01106 16654\n",
      "NA21309 16651\n",
      "HG02148 16651\n",
      "NA19240 16650\n",
      "HG00673 16651\n",
      "HG02818 16651\n",
      "HG02723 16651\n",
      "HG03579 16650\n",
      "HG02630 16653\n",
      "HG01952 16651\n",
      "HG00621 16651\n",
      "HG01358 16651\n",
      "HG02055 16653\n",
      "HG03516 16651\n",
      "HG03492 16652\n",
      "HG01109 16653\n",
      "HG03486 16650\n",
      "HG02145 16653\n",
      "HG01258 16652\n",
      "HG00438 16653\n",
      "HG01978 16652\n",
      "HG00741 16652\n",
      "HG01361 16651\n",
      "HG02486 16651\n"
     ]
    }
   ],
   "source": [
    "samplelist = find_sample_list(spgraph)\n",
    "for sample in samplelist:\n",
    "    subgraph = AGG.SubGraph(spgraph, [sample,  '012920'])\n",
    "    try:\n",
    "        path = AGG.find_most_supported_path(subgraph, [sample])\n",
    "        print(sample, len(path[0]))\n",
    "        sequence = path[0]\n",
    "        outputfile = \"../optimal_path_from_graph/%s.fasta\" % sample\n",
    "        write_fasta(outputfile, sequence, sample)\n",
    "    except:\n",
    "        print(sample, \"Fail\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4ecc8b61",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = ['HG002']\n",
    "subgraph = AGG.SubGraph(spgraph, sample)\n",
    "edgelist = subgraph.edges.keys()\n",
    "for edge in edgelist:\n",
    "    assert len(set(sample) & set(subgraph.edges[edge].get('strain', \"\")))>0, print(subgraph.edges[edge].get('strain', \"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e0e55715",
   "metadata": {},
   "outputs": [],
   "source": [
    "for edge in spgraph.outgoing['A000151']:\n",
    "    if sample in (spgraph.edges[edge]['strain']):\n",
    "        print(edge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "12a2a90d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#AGG.write_gfa(subgraph.anchor, subgraph.edges, \"./HG002_subgraph_k31.gfa\")\n",
    "\n",
    "def find_most_supported_path(graph, sample, ref = '012920'):\n",
    "    src = sorted(graph.anchor.keys())[0]\n",
    "    dst = sorted(graph.anchor.keys())[-1]\n",
    "    Path = []\n",
    "\n",
    "    visited = set()\n",
    "    \n",
    "    def find_most_supported_edge(graph, sample, src, ref):\n",
    "        \n",
    "        edgelist = graph.outgoing.get(src, [])\n",
    "        sample_edge = []\n",
    "        for edge in edgelist:\n",
    "            if sample in graph.edges[edge]['strain']:\n",
    "                sample_edge += [edge]\n",
    "        if len(sample_edge)<1: # no haplotype for this sample at this place the go to the reference one\n",
    "            for ref_edge in edgelist:\n",
    "                if ref in set(graph.edges[ref_edge]['strain']):\n",
    "                    break\n",
    "            edge = ref_edge\n",
    "\n",
    "        elif len(sample_edge) == 1:\n",
    "            edge = sample_edge[0]\n",
    "        else:\n",
    "            # find the most supported edge\n",
    "            read_count = [len(graph.edges[e]['reads']) for e in sample_edge]\n",
    "\n",
    "            index = numpy.argmax(read_count)\n",
    "\n",
    "            edge = sample_edge[index]\n",
    "            \n",
    "        return edge\n",
    "            \n",
    "    while src != dst:\n",
    "        edge = find_most_supported_edge(graph, sample, src, ref)\n",
    "        Path += [src, edge]\n",
    "\n",
    "        visited.add(src)\n",
    "\n",
    "        src = graph.outgoing[edge][0]\n",
    "\n",
    "        assert src.startswith(\"A\") or src == \"SINK\"\n",
    "\n",
    "        if src in visited:\n",
    "            break\n",
    "    \n",
    "    if src == dst:\n",
    "        try:\n",
    "            edge = find_most_supported_edge(graph, sample, src, ref)\n",
    "            Path += [src, edge]\n",
    "        except:\n",
    "            Path = Path\n",
    "        \n",
    "    reconstruct = ''\n",
    "    for item in Path:\n",
    "        if item == \"SOURCE\":\n",
    "            reconstruct += \"\"\n",
    "        elif item.startswith('A'):\n",
    "            reconstruct += graph.anchor[item]['seq']\n",
    "        elif item.startswith(\"E\"):\n",
    "            reconstruct += graph.edges[item]['seq']\n",
    "        else:\n",
    "            print(item)\n",
    "\n",
    "    return reconstruct, Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "af231bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract most supported path\n",
    "path = AGG.find_most_supported_path(subgraph, sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ebd06ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# call major haplotype variants\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f9ae5b11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_variant_position(cigar):\n",
    "    ref_pos = []\n",
    "    alt_pos = []\n",
    "    var_type = []\n",
    "    alt_i = 0\n",
    "    ref_i = 0\n",
    "    for i, s in enumerate(cigar):\n",
    "        if s == 'I':\n",
    "            if ref_i > 0:\n",
    "                ref_pos.append(ref_i-1)\n",
    "            else:\n",
    "                ref_pos.append(ref_i)    \n",
    "            alt_pos.append(alt_i)\n",
    "            var_type.append(\"I\")\n",
    "            \n",
    "            alt_i += 1\n",
    "            \n",
    "        if s == 'D':\n",
    "            ref_pos.append(ref_i)\n",
    "            if alt_i > 0:\n",
    "                alt_pos.append(alt_i-1)\n",
    "            else:\n",
    "                alt_pos.append(alt_i)\n",
    "            var_type.append(\"D\")\n",
    "            ref_i += 1\n",
    "            \n",
    "        if s == 'X':\n",
    "            ref_pos.append(ref_i)\n",
    "            alt_pos.append(alt_i)\n",
    "            alt_i += 1\n",
    "            ref_i += 1\n",
    "            var_type.append(\"X\")\n",
    "            \n",
    "        if s == '=':\n",
    "            alt_i += 1\n",
    "            ref_i += 1\n",
    "\n",
    "    return ref_pos, alt_pos, var_type\n",
    "\n",
    "def findBedge(Graph, src, dst, refstrain, k):\n",
    "    paths = AGG.Find_all_Path_between_anchors(Graph, src, dst, {refstrain})\n",
    "    subpaths = paths.subpath\n",
    "\n",
    "    if len(subpaths) < 1:\n",
    "        return \"\"\n",
    "    for p, strain in subpaths:\n",
    "        seq = AGG.reconstruct_path_seq(spgraph, path = p)\n",
    "    return p[1:-1], seq[k:-k]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a5c934d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for edge in spgraph.outgoing[src]:\n",
    "#     if ref in spgraph.edges[edge]['strain']:\n",
    "#         print(spgraph.edges[edge]['seq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "731da014",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(src, dst)\n",
    "# ref_seq = findBedge(spgraph, src, dst, \"NC_012920\", k)\n",
    "# ref_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "42b38171",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find variants along the path\n",
    "\n",
    "\n",
    "def get_snps(subgraph, mpath, k, ref):\n",
    "    SNPs = {} # Var[refpos]['edgename'] = ['A']\n",
    "    for edge in mpath:\n",
    "        if edge.startswith('E'):\n",
    "            cigar = subgraph.edges[edge].get('variants', \"\")\n",
    "            if cigar == \"\":\n",
    "                continue\n",
    "            src = subgraph.incoming[edge][0]\n",
    "            if src != \"SOURCE\":\n",
    "                refstart = int(subgraph.anchor[src][\"pos\"]) + k\n",
    "            else:\n",
    "                raise \"SOURCE or SINK node\"\n",
    "\n",
    "            # find reference seq\n",
    "            dst = subgraph.outgoing[edge][0]\n",
    "            ref_edge_list, ref_seq = findBedge(spgraph, src, dst, \"NC_012920\", k)\n",
    "            ref_edge = ref_edge_list[0]\n",
    "            expanded_cigar = AGG.processCigar(cigar)\n",
    "            refpos, altpos, var_type = get_variant_position(AGG.processCigar(cigar))\n",
    "            alt_seq = subgraph.edges[edge]['seq']\n",
    "            # only record SNPs\n",
    "            for i, vart in enumerate(var_type):\n",
    "                if vart == \"X\":\n",
    "                    rp = refpos[i]\n",
    "                    ap = altpos[i]\n",
    "                    pos = refstart + rp\n",
    "                    SNPs[pos] = SNPs.get(pos, {})\n",
    "                    SNPs[pos][edge] = SNPs[pos].get(edge, {})                \n",
    "                    SNPs[pos][edge]['base'] = SNPs[pos][edge].get('base', \"\") + alt_seq[ap]\n",
    "                    SNPs[pos][edge]['altoffset'] = SNPs[pos][edge].get('altoffset', []) + [ap]\n",
    "                    SNPs[pos][ref_edge] = SNPs[pos].get(ref_edge, {})\n",
    "                    SNPs[pos][ref_edge]['base'] = ref_seq[refpos[i]]\n",
    "                    SNPs[pos]['refbase'] = SNPs[pos][ref_edge]['base']\n",
    "    return SNPs\n",
    "#     node = graph.outgoing[edge][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "53917ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def find_allVar(graph, refstrain, founder, k):\n",
    "#     #print cigar, refpos, altpos\n",
    "#     return Var\n",
    "\n",
    "# # Single Sample VCF\n",
    "\n",
    "def get_VCF_file(Var, samplelist):\n",
    "    Mat = {}\n",
    "\n",
    "    for pos, D in Var.items():\n",
    "        Mat[pos] = {}\n",
    "        for edge, B in D.items():\n",
    "            if edge == \"refbase\":\n",
    "                continue\n",
    "            for strain in samplelist:\n",
    "                Mat[pos][strain] = Mat[pos].get(strain, set()) | set([B['base']])\n",
    "    # write VCF\n",
    "    VCF = {}\n",
    "    for pos, D in Mat.items():\n",
    "        refbase = Var[pos]['refbase']\n",
    "        \n",
    "        if list(D.values())[0] == set([refbase]):\n",
    "            continue    \n",
    "        VCF[pos]= VCF.get(pos, {})\n",
    "        for strain, base in D.items():\n",
    "            VCF[pos][\"CHROM\"] = \"chrM\"\n",
    "            VCF[pos][\"POS\"] = pos\n",
    "            VCF[pos]['ID'] = \".\"\n",
    "            VCF[pos]['REF'] = refbase\n",
    "            #print(base - set([refbase]))\n",
    "            VCF[pos]['ALT'] = \",\".join(base - set([refbase]))\n",
    "            VCF[pos]['QUAL'] = \".\"\n",
    "            VCF[pos]['FILTER'] = \".\"\n",
    "            VCF[pos]['INFO'] = \".\"# Indels[pos][edge]['info'] = \"INS\"\n",
    "            VCF[pos]['FORMAT'] = \"GT\"\n",
    "            #print(base)\n",
    "            # allelelist = []\n",
    "            # if refbase in base:\n",
    "            #     allelelist += [0]\n",
    "\n",
    "            # for i,s in enumerate(base - set([refbase])):\n",
    "            #     allelelist += [i+1]\n",
    "            # VCF[pos][strain] = \"/\".join([str(item) for item in allelelist])\n",
    "            VCF[pos][strain] = \"/\".join([str(item) for item in base])\n",
    "    return VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "331e1297",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HG002']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "03ce3143",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unhashable type: 'list'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/ws/928zzdt94qn3749jx20228j80000gq/T/ipykernel_13841/2034287089.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"012920\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mSNPs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_snps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubgraph\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmpath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mVCF_SNPS\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_VCF_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSNPs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0msample\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mVCF_SNPS\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/ws/928zzdt94qn3749jx20228j80000gq/T/ipykernel_13841/109387434.py\u001b[0m in \u001b[0;36mget_VCF_file\u001b[0;34m(Var, samplelist)\u001b[0m\n\u001b[1;32m     14\u001b[0m                 \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mstrain\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msamplelist\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m                 \u001b[0mMat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpos\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstrain\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mMat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpos\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstrain\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mB\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'base'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m     \u001b[0;31m# write VCF\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m     \u001b[0mVCF\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: unhashable type: 'list'"
     ]
    }
   ],
   "source": [
    "mpath = path[1]\n",
    "k = 31\n",
    "ref = \"012920\"\n",
    "SNPs = get_snps(subgraph, mpath, k, ref)\n",
    "VCF_SNPS = get_VCF_file(SNPs, [sample])\n",
    "df = pd.DataFrame(VCF_SNPS).T\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e130dbe9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "31386b4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NA18906 (80, 10)\n",
      "HG03453 (49, 10)\n",
      "HG00735 (39, 10)\n",
      "HG02109 (82, 10)\n",
      "012920 (0, 0)\n",
      "HG02572 (77, 10)\n",
      "HG02257 (39, 10)\n",
      "HG02717 (55, 10)\n",
      "HG03540 (55, 10)\n",
      "HG00733 (28, 10)\n",
      "HG01123 (27, 10)\n",
      "HG02886 (48, 10)\n",
      "HG01071 (27, 10)\n",
      "HG002 (12, 10)\n",
      "HG01243 (77, 10)\n",
      "HG02559 (21, 10)\n",
      "NA20129 (29, 10)\n",
      "HG01928 (24, 10)\n",
      "HG03098 (50, 10)\n",
      "HG02622 (30, 10)\n",
      "HG02080 (34, 10)\n",
      "HG01891 (54, 10)\n",
      "HG01175 (29, 10)\n",
      "HG01106 (34, 10)\n",
      "NA21309 (52, 10)\n",
      "HG02148 (30, 10)\n",
      "NA19240 (35, 10)\n",
      "HG00673 (29, 10)\n",
      "HG02818 (38, 10)\n",
      "HG02723 (40, 10)\n",
      "HG03579 (55, 10)\n",
      "HG02630 (53, 10)\n",
      "HG01952 (32, 10)\n",
      "HG00621 (20, 10)\n",
      "HG01358 (32, 10)\n",
      "HG02055 (34, 10)\n",
      "HG03516 (49, 10)\n",
      "HG03492 (29, 10)\n",
      "HG01109 (82, 10)\n",
      "HG03486 (52, 10)\n",
      "HG02145 (54, 10)\n",
      "HG01258 (26, 10)\n",
      "HG00438 (31, 10)\n",
      "HG01978 (5, 10)\n",
      "HG00741 (79, 10)\n",
      "HG01361 (27, 10)\n",
      "HG02486 (10, 10)\n"
     ]
    }
   ],
   "source": [
    "# main function for calling major SNPs\n",
    "k = 31\n",
    "refstrain = \"012920\"\n",
    "# samplelist = ['HG00438', \"HG00621\"]\n",
    "#samplelist = [\"HG002\",'HG00621', \"HG00438\"]\n",
    "for sample in samplelist:\n",
    "    subgraph = AGG.SubGraph(spgraph, [sample,  '012920']) # always keep the reference\n",
    "    path = AGG.find_most_supported_path(subgraph, sample)\n",
    "    mpath = path[1]\n",
    "    SNPs = get_snps(subgraph, mpath, k, ref)\n",
    "    VCF_SNPS = get_VCF_file(SNPs, [sample])\n",
    "    df = pd.DataFrame(VCF_SNPS).T\n",
    "    print(sample, df.shape)\n",
    "    #df.to_csv(\"../optimal_path_from_graph/%s.snps.vcf\" % sample, index = False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9dc214d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_indels(subgraph, mpath, k, ref):\n",
    "    Indels = {} # Var[refpos]['edgename'] = ['A']\n",
    "    for edge in mpath:\n",
    "        if edge.startswith('E'):\n",
    "            cigar = subgraph.edges[edge].get('variants', \"\")\n",
    "            if cigar == \"\":\n",
    "                continue\n",
    "            src = subgraph.incoming[edge][0]\n",
    "            if src != \"SOURCE\":\n",
    "                refstart = int(subgraph.anchor[src][\"pos\"]) + k\n",
    "            else:\n",
    "                raise \"SOURCE or SINK node\"\n",
    "\n",
    "            # find reference seq\n",
    "            dst = subgraph.outgoing[edge][0]\n",
    "            ref_edge_list, ref_seq = findBedge(spgraph, src, dst, \"NC_012920\", k)\n",
    "            ref_edge = ref_edge_list[0]\n",
    "            expanded_cigar = AGG.processCigar(cigar)\n",
    "            refpos, altpos, var_type = get_variant_position(AGG.processCigar(cigar))\n",
    "            alt_seq = subgraph.edges[edge]['seq']\n",
    "            # only record SNPs\n",
    "            for i, vart in enumerate(var_type):\n",
    "                if vart == \"I\":\n",
    "                    rp = refpos[i]\n",
    "                    ap = altpos[i]\n",
    "                    pos = refstart + rp\n",
    "                    Indels[pos] = Indels.get(pos, {})\n",
    "                    Indels[pos][edge] = Indels[pos].get(edge, {})                \n",
    "                    Indels[pos][edge]['base'] = Indels[pos][edge].get('base', \"\") + alt_seq[ap:ap+2]\n",
    "                    Indels[pos][edge]['info'] = \"INS\"\n",
    "                    Indels[pos][edge]['altoffset'] = Indels[pos][edge].get('altoffset', []) + [ap]\n",
    "                    Indels[pos][ref_edge] = Indels[pos].get(ref_edge, {})\n",
    "                    Indels[pos][ref_edge]['base'] = Indels[pos][ref_edge].get(\"base\", \"\") + ref_seq[rp-1]\n",
    "                    Indels[pos]['refbase'] = Indels[pos][ref_edge]['base']\n",
    "\n",
    "                if vart == \"D\":\n",
    "                    rp = refpos[i]\n",
    "                    ap = altpos[i]\n",
    "                    pos = refstart + rp\n",
    "                    Indels[pos] = Indels.get(pos, {})\n",
    "                    Indels[pos][edge] = Indels[pos].get(edge, {})                \n",
    "                    Indels[pos][edge]['base'] = Indels[pos][edge].get('base', \"\") + alt_seq[ap]\n",
    "                    Indels[pos][edge]['info'] = \"DEL\"\n",
    "                    Indels[pos][edge]['altoffset'] = Indels[pos][edge].get('altoffset', []) + [ap-1]\n",
    "                    Indels[pos][ref_edge] = Indels[pos].get(ref_edge, {})\n",
    "                    Indels[pos][ref_edge]['base'] = Indels[pos][ref_edge].get(\"base\", \"\") + ref_seq[rp-1:rp+1]\n",
    "                    Indels[pos]['refbase'] = Indels[pos][ref_edge]['base']\n",
    "    return Indels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "da3016e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG002 (3, 10)\n"
     ]
    }
   ],
   "source": [
    "k = 31\n",
    "refstrain = \"012920\"\n",
    "# samplelist = ['HG00438', \"HG00621\"]\n",
    "#samplelist = [\"HG002\",'HG00621', \"HG00438\"]\n",
    "for sample in [\"HG002\"]:\n",
    "    subgraph = AGG.SubGraph(spgraph, [sample,  '012920']) # always keep the reference\n",
    "    path = AGG.find_most_supported_path(subgraph, sample)\n",
    "    mpath = path[1]\n",
    "    Indels = get_indels(subgraph, mpath, k, ref)\n",
    "    VCF_SNPS = get_VCF_file(Indels, [sample])\n",
    "    df = pd.DataFrame(VCF_SNPS).T\n",
    "    print(sample, df.shape)\n",
    "    #df.to_csv(\"../optimal_path_from_graph/%s.indels.vcf\" % sample, index = False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c0277822",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>HG002</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>chrM</td>\n",
       "      <td>309</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>C/CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>chrM</td>\n",
       "      <td>310</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>C/CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3107</th>\n",
       "      <td>chrM</td>\n",
       "      <td>3107</td>\n",
       "      <td>.</td>\n",
       "      <td>CN</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>CN/C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM   POS ID REF ALT QUAL FILTER INFO FORMAT HG002\n",
       "309   chrM   309  .   C  CT    .      .    .     GT  C/CT\n",
       "310   chrM   310  .   C  CC    .      .    .     GT  C/CC\n",
       "3107  chrM  3107  .  CN   C    .      .    .     GT  CN/C"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef599e8",
   "metadata": {},
   "source": [
    "# call major and heteroplasmy from the whole subgraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7c2b9966",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.75"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "15 * 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "aaf366ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def prune_under_supported_path(spgraph, samplelist, threshold = 0.05):\n",
    "#     for sample in samplelist:\n",
    "#         sub_spgraph = AGG.SubGraph(spgraph, [sample])\n",
    "#         nodelist = sorted(sub_spgraph.outgoing.keys())    \n",
    "#         readsets = AGG.find_all_reads(sub_spgraph)\n",
    "#         readnum = sum([1 for r in list(readsets) if r.endswith(sample)])\n",
    "#         print(readnum)\n",
    "#         for i in range(len(nodelist)-1):\n",
    "#             sanchor = nodelist[i]\n",
    "#             eanchor = nodelist[i+1]\n",
    "#             paths = sub_spgraph.outgoing[sanchor]\n",
    "#             supported = [p for p in paths if p.startswith('E') if len(set(sub_spgraph.edges[p]['reads'])) > readnum * threshold]\n",
    "#             unsupported = [p for p in paths if p not in supported]\n",
    "#             # delete unsupported\n",
    "#             for e in unsupported:\n",
    "#                 if e.startswith(\"E\"):\n",
    "#                     del sub_spgraph.edges[e]\n",
    "\n",
    "#     return sub_spgraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "fe0d6aab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG002 (12, 10) (131, 10)\n"
     ]
    }
   ],
   "source": [
    "threshold = 0.05\n",
    "k = 31\n",
    "refstrain = \"012920\"\n",
    "# samplelist = ['HG00438', \"HG00621\"]\n",
    "#samplelist = [\"HG002\",'HG00621', \"HG00438\"]\n",
    "for sample in ['HG002']:\n",
    "    subgraph = AGG.SubGraph(spgraph, [sample,  '012920']) # always keep the reference\n",
    "    readnum = [read for read in AGG.find_all_reads(subgraph) if read.endswith(sample)]\n",
    "    \n",
    "    #sub_graph = prune_under_supported_path(subgraph, [sample,  '012920'], threshold = 0.005)\n",
    "    #path = AGG.find_most_supported_path(subgraph, sample)\n",
    "    paths = list(subgraph.edges.keys())\n",
    "    mpath = [edge for edge in paths if len([r for r in subgraph.edges[edge]['reads'] if r.endswith(sample)]) > len(readnum) * threshold]\n",
    "    SNPs = get_snps(subgraph, mpath, k, ref)\n",
    "    Indels = get_indels(subgraph, mpath, k, ref)\n",
    "    VCF_SNPS = get_VCF_file(SNPs, [sample])\n",
    "    VCF_Indels = get_VCF_file(Indels, [sample])\n",
    "    df = pd.DataFrame(VCF_SNPS).T\n",
    "    di = pd.DataFrame(VCF_Indels).T\n",
    "    print(sample, df.shape, di.shape)\n",
    "    #df.to_csv(\"../optimal_path_from_graph/%s.indels.vcf\" % sample, index = False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "9bd04202",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52=\n",
      "125=1X46=2I1=1I56=\n",
      "125=1X46=1I1=1I56=\n",
      "125=1X46=2I1=2I56=\n",
      "125=1X47=1I56=\n",
      "125=1X46=3I1=1I56=\n",
      "58=1X1=\n",
      "124=\n",
      "106=1X82=\n",
      "106=1X51=1D30=\n",
      "64=1D41=1X82=\n",
      "94=\n",
      "39=1I55=\n",
      "94=1I\n",
      "34=1D59=\n",
      "10=1D83=\n",
      "98=\n",
      "9=1D88=\n",
      "66=\n",
      "51=1I15=\n",
      "10=1I56=\n",
      "108=\n",
      "24=1D83=\n",
      "11=1D96=\n",
      "84=1X48=\n",
      "70=\n",
      "26=1I44=\n",
      "135=\n",
      "67=1I68=\n",
      "94=1D40=\n",
      "107=1D27=\n",
      "84=\n",
      "1=1D82=\n",
      "59=\n",
      "71=\n",
      "59=1D11=\n",
      "38=\n",
      "19=1I19=\n",
      "100=\n",
      "58=1D41=\n",
      "62=\n",
      "40=1D21=\n",
      "76=\n",
      "31=1I45=\n",
      "79=\n",
      "53=1I26=\n",
      "72=\n",
      "95=\n",
      "65=1I30=\n",
      "89=1D5=\n",
      "77=\n",
      "74=1D2=\n",
      "89=\n",
      "48=1D44=\n",
      "19=1D28=1D44=\n",
      "22=1I26=1D44=\n",
      "79=\n",
      "51=1D27=\n",
      "67=\n",
      "204=\n",
      "181=1I23=\n",
      "188=1I16=\n",
      "175=1I29=\n",
      "118=\n",
      "21=1I97=\n",
      "53=\n",
      "41=1D11=\n",
      "16=1I37=\n",
      "116=\n",
      "64=\n",
      "15=1I49=\n",
      "166=\n",
      "40=1D125=\n",
      "38=1X57=\n",
      "38=1X29=1I28=\n",
      "38=1X29=1D27=\n",
      "108=\n",
      "38=1D69=\n",
      "14=1D93=\n",
      "103=\n",
      "71=1X279=\n",
      "71=1X277=1I2=\n",
      "71=1X155=1D123=\n",
      "71=1X100=1I179=\n",
      "47=1D23=1X279=\n",
      "27=1I44=1X155=1I62=1D61=\n",
      "86=\n",
      "72=1I14=\n",
      "137=\n",
      "90=1D46=\n",
      "39=1D97=\n",
      "106=\n",
      "76=1I30=\n",
      "46=1D59=\n",
      "231=\n",
      "74=1D156=\n",
      "90=\n",
      "70=1D19=\n",
      "78=\n",
      "31=1D46=\n",
      "28=1I50=\n",
      "72=\n",
      "127=\n",
      "76=1D50=\n",
      "93=1I34=\n",
      "92=1D34=\n",
      "152=\n",
      "149=\n",
      "134=1D14=\n",
      "13=1I115=1I21=\n",
      "64=1D84=\n",
      "46=\n",
      "15=1D30=\n",
      "60=\n",
      "15=1I45=\n",
      "14=1X117=\n",
      "260=\n",
      "244=1D15=\n",
      "196=\n",
      "110=1D85=\n",
      "35=1I148=1I13=\n",
      "63=\n",
      "46=1D1=1D14=\n",
      "52=\n",
      "166=\n",
      "89=1I77=\n",
      "125=\n",
      "23=1D101=\n",
      "28=1I97=\n",
      "118=\n",
      "116=1D1=\n",
      "86=1D31=\n",
      "23=1I95=\n",
      "139=\n",
      "138=1D\n",
      "130=1D8=\n",
      "94=1D20=\n",
      "115=\n",
      "95=1I20=\n",
      "95=1X28=\n",
      "46=\n",
      "67=\n",
      "58=1D8=\n",
      "66=1D\n",
      "67=1I1=1X33=\n",
      "68=1X33=\n",
      "68=1X29=1I4=\n",
      "65=\n",
      "88=\n",
      "154=\n",
      "127=1D26=\n",
      "118=1D35=\n",
      "10=1I105=1D38=\n",
      "91=\n",
      "91=1I\n",
      "38=1I151=\n",
      "189=\n",
      "128=1I61=\n",
      "155=\n",
      "134=1I21=\n",
      "3=1D151=\n",
      "44=1I75=\n",
      "119=\n",
      "100=1I19=\n",
      "127=\n",
      "50=1D76=\n",
      "20=1D106=\n",
      "153=\n",
      "116=1I37=\n",
      "59=1D93=\n",
      "57=\n",
      "15=1I74=\n",
      "89=\n",
      "136=\n",
      "136=1I\n",
      "102=\n",
      "99=1D2=\n",
      "113=\n",
      "99=1D13=\n",
      "91=1I22=\n",
      "33=1I80=\n",
      "32=1D80=\n",
      "93=\n",
      "42=1I51=\n",
      "63=\n",
      "58=\n",
      "70=\n",
      "105=\n",
      "55=1I50=\n",
      "57=\n",
      "59=\n",
      "56=\n",
      "25=1D115=\n",
      "141=\n",
      "67=1I74=\n",
      "26=1I115=\n",
      "116=\n",
      "42=1D73=\n",
      "128=\n",
      "27=1I99=\n",
      "126=\n",
      "30=1D95=\n",
      "79=\n",
      "88=\n",
      "17=1D70=\n",
      "119=\n",
      "131=\n",
      "88=\n",
      "46=\n",
      "46=1I57=\n",
      "103=\n",
      "53=1D14=\n",
      "68=\n",
      "59=1D8=\n",
      "53=\n",
      "58=\n",
      "4=1D53=\n",
      "107=1D\n",
      "108=\n",
      "91=\n",
      "177=\n",
      "95=1I82=\n",
      "94=\n",
      "173=\n",
      "119=1D53=\n",
      "106=1D66=\n",
      "185=\n",
      "47=1D74=1I63=\n",
      "136=\n",
      "65=\n",
      "52=1D12=\n",
      "109=\n",
      "12=1D96=\n",
      "137=\n",
      "6=1X150=1X19=\n",
      "6=1X90=1I60=1X19=\n",
      "317=\n",
      "305=1D11=\n",
      "163=1D153=\n",
      "35=1I282=\n",
      "126=\n",
      "104=\n",
      "2=1D101=\n",
      "22=1I82=\n",
      "287=1X90=\n",
      "5=\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'variants'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/ws/928zzdt94qn3749jx20228j80000gq/T/ipykernel_13841/3870660570.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0medge\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmpath\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubgraph\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0medges\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0medge\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'variants'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 'variants'"
     ]
    }
   ],
   "source": [
    "for edge in mpath:\n",
    "    print(subgraph.edges[edge]['variants'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "cf2ed23a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>HG002</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>chrM</td>\n",
       "      <td>309</td>\n",
       "      <td>.</td>\n",
       "      <td>CCCCCCCC</td>\n",
       "      <td>CCCT,CT,CCCCCT</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>CCCCCT/CCCT/CT/CCCCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>chrM</td>\n",
       "      <td>310</td>\n",
       "      <td>.</td>\n",
       "      <td>CCCCCC</td>\n",
       "      <td>CCCC,CC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>CCCC/CC/CCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>802</th>\n",
       "      <td>chrM</td>\n",
       "      <td>802</td>\n",
       "      <td>.</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>AC/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>chrM</td>\n",
       "      <td>708</td>\n",
       "      <td>.</td>\n",
       "      <td>CC</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>CC/C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>902</th>\n",
       "      <td>chrM</td>\n",
       "      <td>902</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GT</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>G/GT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15682</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15682</td>\n",
       "      <td>.</td>\n",
       "      <td>AA</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>AA/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15540</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15540</td>\n",
       "      <td>.</td>\n",
       "      <td>CC</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>CC/C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15411</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15411</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>C/CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15884</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15884</td>\n",
       "      <td>.</td>\n",
       "      <td>GG</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>G/GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15903</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15903</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>AC/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>131 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM    POS ID       REF             ALT QUAL FILTER INFO FORMAT  \\\n",
       "309    chrM    309  .  CCCCCCCC  CCCT,CT,CCCCCT    .      .    .     GT   \n",
       "310    chrM    310  .    CCCCCC         CCCC,CC    .      .    .     GT   \n",
       "802    chrM    802  .        AC               A    .      .    .     GT   \n",
       "708    chrM    708  .        CC               C    .      .    .     GT   \n",
       "902    chrM    902  .         G              GT    .      .    .     GT   \n",
       "...     ...    ... ..       ...             ...  ...    ...  ...    ...   \n",
       "15682  chrM  15682  .        AA               A    .      .    .     GT   \n",
       "15540  chrM  15540  .        CC               C    .      .    .     GT   \n",
       "15411  chrM  15411  .         C              CT    .      .    .     GT   \n",
       "15884  chrM  15884  .        GG               G    .      .    .     GT   \n",
       "15903  chrM  15903  .         A              AC    .      .    .     GT   \n",
       "\n",
       "                         HG002  \n",
       "309    CCCCCT/CCCT/CT/CCCCCCCC  \n",
       "310             CCCC/CC/CCCCCC  \n",
       "802                       AC/A  \n",
       "708                       CC/C  \n",
       "902                       G/GT  \n",
       "...                        ...  \n",
       "15682                     AA/A  \n",
       "15540                     CC/C  \n",
       "15411                     C/CT  \n",
       "15884                     G/GG  \n",
       "15903                     AC/A  \n",
       "\n",
       "[131 rows x 10 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "di"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "72c7e5de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# traverse the graph\n",
    "# from chat-gpt\n",
    "# def dfs(graph, start, end, reads):\n",
    "#     stack = [(start, [start], reads)]\n",
    "#     while stack:\n",
    "#         (node, path, reads) = stack.pop()\n",
    "#         for next_node, next_reads in graph[node].items():\n",
    "#             intersect_reads = reads.intersection(next_reads)\n",
    "#             if not intersect_reads:\n",
    "#                 continue\n",
    "#             if next_node == end:\n",
    "#                 yield path + [next_node]\n",
    "#             else:\n",
    "#                 stack.append((next_node, path + [next_node], intersect_reads))\n",
    "\n",
    "# # Example usage:\n",
    "# graph = {\n",
    "#     'A': {'B': {'read1', 'read2'}, 'C': {'read2', 'read3'}},\n",
    "#     'B': {'A': {'read1'}, 'D': {'read4'}, 'E': {'read5'}},\n",
    "#     'C': {'A': {'read2'}, 'F': {'read3'}},\n",
    "#     'D': {'B': {'read4'}},\n",
    "#     'E': {'B': {'read5'}, 'F': {'read6'}},\n",
    "#     'F': {'C': {'read3'}, 'E': {'read6'}},\n",
    "# }\n",
    "# start = 'A'\n",
    "# end = 'F'\n",
    "# reads = {'read1', 'read2', 'read3'}\n",
    "\n",
    "# paths = list(dfs(graph, start, end, reads))\n",
    "# for path in paths:\n",
    "#     print(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fe58fca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55a3e16a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def dfs(graph, node, end, path, reads):\n",
    "#     path = path + [node]\n",
    "#     if node == end:\n",
    "#         return [path]\n",
    "#     paths = []\n",
    "#     for next_node, next_reads in graph[node].items():\n",
    "#         intersect_reads = reads.intersection(next_reads)\n",
    "#         if not intersect_reads or next_node in path:\n",
    "#             continue\n",
    "#         newpaths = dfs(graph, next_node, end, path, intersect_reads)\n",
    "#         for newpath in newpaths:\n",
    "#             paths.append(newpath)\n",
    "#     return paths\n",
    "\n",
    "# # Example usage:\n",
    "# graph = {\n",
    "#     'A': {'B': {'read1', 'read2'}, 'C': {'read2', 'read3'}},\n",
    "#     'B': {'A': {'read1'}, 'D': {'read4'}, 'E': {'read5'}},\n",
    "#     'C': {'A': {'read2'}, 'F': {'read3'}},\n",
    "#     'D': {'B': {'read4'}},\n",
    "#     'E': {'B': {'read5'}, 'F': {'read6'}},\n",
    "#     'F': {'C': {'read3'}, 'E': {'read6'}},\n",
    "# }\n",
    "# start = 'A'\n",
    "# end = 'F'\n",
    "# reads = {'read1', 'read2', 'read3'}\n",
    "\n",
    "# paths = dfs(graph, start, end, [], reads)\n",
    "# for path in paths:\n",
    "#     print(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0314987e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
